### PR TITLE
Additional flags for xcodebuild

### PR DESCRIPTION
**Pull request by RusKryzhanovskiy**
## Short description
I added two parameters to the `xcodebuild` command which is invoked by the `flutter build ipa` command: 

- -allowProvisioningUpdates

Allow xcodebuild to communicate with the Apple Developer website. For automatically signed targets, xcodebuild will create and update profiles, app IDs, and certificates. For manually signed targets, xcodebuild will download missing or updated provisioning profiles. Requires a developer account to have been added in Xcode's Accounts preference pane.

- -allowProvisioningDeviceRegistration

Allow xcodebuild to register your destination device on the developer portal if necessary. This flag only takes effect if -allowProvisioningUpdates is also passed.

## *List which issues are fixed by this PR.*
https://github.com/flutter/flutter/issues/83125

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jmagman** [comment]
The build_ipa_test tests need to be updated:
```
07:46 +390: test/commands.shard/hermetic/build_ipa_test.dart: ipa build invokes xcode build export archive                                                                                             
Expected: [
            'xcrun',
            'xcodebuild',
            '-exportArchive',
            '-archivePath',
            '/build/ios/archive/Runner.xcarchive',
            '-exportPath',
            '/build/ios/ipa',
            '-exportOptionsPlist',
            '/ExportOptions.plist'
          ]
  Actual: [
            'xcrun',
            'xcodebuild',
            '-exportArchive',
            '-allowProvisioningDeviceRegistration',
            '-allowProvisioningUpdates',
            '-archivePath',
            '/build/ios/archive/Runner.xcarchive',
            '-exportPath',
            '/build/ios/ipa',
            '-exportOptionsPlist',
            '/ExportOptions.plist'
          ]
   Which: at location [3] is '-allowProvisioningDeviceRegistration' instead of '-archivePath'
```
https://github.com/flutter/flutter/blob/de23eb7b292a5338d4133df986f8e8c4f5db8b6c/packages/flutter_tools/test/commands.shard/hermetic/build_ipa_test.dart#L107-L119

**Comment by RusKryzhanovskiy** [comment]
@jmagman should I update this section only? 
Screenshot 2021-05-29 at 17 11 09

**Comment by YehorHolyk** [comment]
Damn, this is amazing. I've been looking for this for a whole year. Thanks for working on this, go ahead, I'm waiting for a new release.

**Comment by YehorHolyk** [comment]
I've been looking for this for a whole month. Thanks for working on this, go ahead, I'm waiting for a new release.

**Comment by RusKryzhanovskiy** [comment]
@jmagman I set these two commands as optional. Now you can use it in different ways:

 `flutter build ipa`,
 `flutter build ipa --allowProvisioningDeviceRegistration --allowProvisioningUpdates` and so on.
 
 Could you check it?

**Review by jmagman — CHANGES_REQUESTED** [review]
1. This needs tests.
https://github.com/flutter/flutter/pull/83128#issuecomment-847341596

2. Since archiving implies codesigning, why not always pass in those flags? 
 This is what's done for `flutter build ios`/`flutter run` when codesigning:
https://github.com/flutter/flutter/blob/3e6fe491f3689973f61f62ceba60908b6c03e14f/packages/flutter_tools/lib/src/ios/mac.dart#L222-L228

I'm not sure why `module_test_ios` is failing with codesigning `EXC_BAD_ACCESS`
[Host_2021-05-30-104058_2FD25848-AC4D-44BA-8A85-4A6C1E1E7EEB.crash.txt](https://github.com/flutter/flutter/files/6578433/Host_2021-05-30-104058_2FD25848-AC4D-44BA-8A85-4A6C1E1E7EEB.crash.txt), it doesn't look like your change would impact that test.  If it fails again on your next update I'll run locally and investigate.

**Comment by dim-apparence** [comment]
Nice works!
Thanks for the PR :)

**Comment by RusKryzhanovskiy** [comment]
@jmagman module_test_ios was failed for some reason.
Could you also review my code again?

**Comment by jmagman** [comment]
> @jmagman module_test_ios was failed for some reason.
> Could you also review my code again?

I couldn't reproduce when I rebased your branch on master.  Could you rebase to confirm these tests are already fixed?

**Review by jmagman — APPROVED** [review]
LGTM once it's rebased on master.

File: packages/flutter_tools/lib/src/commands/build_ios.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/commands/build_ios.dart:150]: Usually we combine into one if condition:
```suggestion
          if (shouldCodesign) ...<String>[
            '-allowProvisioningDeviceRegistration',
            '-allowProvisioningUpdates',
          ]
```

**Comment by RusKryzhanovskiy** [comment]
@jmagman there is a command list I executed in my repo.

git remote add upstream https://github.com/flutter/flutter.git
git fetch upstream
git rebase upstream/master
git push origin master --force

This helped me. Thank you so much! 
Are any extra actions required on my part?

**Comment by jmagman** [comment]
Will merge this when the tests pass.  Thank you for your PR!

**Comment by RusKryzhanovskiy** [comment]
@jmagman flutter_tree test changed it's status for some reason. What should I do?

**Comment by jmagman** [comment]
> @jmagman flutter_tree test changed it's status for some reason. What should I do?

That isn't related to your commit, it means that there's a test failure in the currently submitted code on top of tree (dashboard at https://flutter-dashboard.appspot.com/#/build).  We don't merge new commits until that's resolved.  So sit tight, it will merge.

**Comment by dim-apparence** [comment]
Hello, is this feature available on stable version ?

**Comment by jmagman** [comment]
It's available in 2.3.0-24.0.pre.

**Comment by thipokch** [comment]
Is this available for Flutter 3?

**Pull request merged**
